File: lib/src/main/java/com/asyncanimator/launcher/manager/OplusAnimManager.java
package com.asyncanimator.launcher.manager;

import com.asyncanimator.launcher.controller.AnimationController;
import com.asyncanimator.launcher.controller.DefaultAnimationController;
import com.asyncanimator.launcher.feature.AnimationFeatureHelper;
import com.asyncanimator.launcher.seq.AnimationSeqHelper;
import com.asyncanimator.launcher.seq.DefaultAnimationSeqHelper;

/**
 * OplusAnimManager — feature flag 驱动的工厂单例。
 *
 * <p>对应分析文档 §6.10。默认所有 helper 返回 {@code Default*BaseClass}（no-op），
 * 当 {@link #supportInterruption()} 为 true 时切换到 {@code Impl}。
 *
 * <p>业务统一通过 {@code OplusAnimManager.getInstance().getAnimController()} 等获取实例，
 * 不需要知道当前返回的是 Default 还是 Impl。
 */
public class OplusAnimManager {

    private static final OplusAnimManager INSTANCE = new OplusAnimManager();

    // 简化版：直接 lazy 创建（生产环境应该是 t4.b 类型懒加载）
    private AnimationController animationController;
    private AnimationSeqHelper animationSeqHelper;

    private OplusAnimManager() {
        if (supportInterruption()) {
            animationController = new AnimationController();
            animationSeqHelper = new AnimationSeqHelper();
        } else {
            animationController = null;
            animationSeqHelper = null;
        }
    }

    public static OplusAnimManager getInstance() {
        return INSTANCE;
    }

    /**
     * feature 开关。简化：默认 true。
     * 生产代码会检查 LauncherAnimConfig / TaskAnimationManager.ENABLE_SHELL_TRANSITIONS
     * 等多个条件。
     */
    public boolean supportInterruption() {
        // demo 默认打开所有功能
        return true;
    }

    public DefaultAnimationController getAnimController() {
        return animationController != null ? animationController : new DefaultAnimationController();
    }

    public DefaultAnimationSeqHelper getAnimationSeqHelper() {
        return animationSeqHelper != null ? animationSeqHelper : new DefaultAnimationSeqHelper();
    }

    public AnimationFeatureHelper getFeatureHelper() {
        return AnimationFeatureHelper.getInstance();
    }

    public void cleanUpRecentsAnimation() {
        if (animationController != null) animationController.cleanUpRecentsAnim();
    }

    /** 重置为 no-op（feature toggle 关闭）。demo 用来演示降级。 */
    public void setInterruptionEnabled(boolean enabled) {
        if (enabled) {
            if (animationController == null) animationController = new AnimationController();
            if (animationSeqHelper == null) animationSeqHelper = new AnimationSeqHelper();
        } else {
            animationController = null;
            animationSeqHelper = null;
        }
    }
}

File: lib/src/main/java/com/asyncanimator/manager/OplusAnimManager.kt
package com.asyncanimator.manager

import com.asyncanimator.api.PublicApi
import com.asyncanimator.control.AnimationController
import com.asyncanimator.control.DefaultAnimationController
import com.asyncanimator.control.checkControllerMainThread
import com.asyncanimator.seq.AnimationSeqHelper
import com.asyncanimator.seq.DefaultAnimationSeqHelper

/**
 * 根据本地开关提供转场控制器与序列协调器的进程内工厂。
 * 首次对象初始化时创建实际实现，关闭后读取返回新的降级实例，再启用时创建新的实际实现。
 * 配置容器的数据变更不会自动切换或重建本工厂，类型名称与现有调用接口保持不变。
 */
@PublicApi
object OplusAnimManager {

    @Volatile
    private var animationControllerImpl: AnimationController? = null
    @Volatile
    private var animationSeqHelperImpl: AnimationSeqHelper? = null

    /**
     * 在单例对象首次初始化时按能力声明建立控制器和序列实例，不是每个 getter 各自惰性构造。
     */
    init {
        if (supportInterruption()) {
            animationControllerImpl = AnimationController()
            animationSeqHelperImpl = AnimationSeqHelper()
        }
    }

    /**
     * 返回本库声明的中断能力占位值，当前固定为 true。
     * 不读取本地工厂开关、系统设置或远程配置；是否返回实际控制器应查询 interruptionEnabled。
     */
    @PublicApi
    fun supportInterruption(): Boolean = true

    /**
     * 读取当前实际控制器；关闭时每次返回新的降级实例，不缓存或共享降级监听注册表。
     */
    @PublicApi
    val animController: DefaultAnimationController
        get() = animationControllerImpl ?: DefaultAnimationController()

    /**
     * 读取当前实际序列 helper；关闭时返回新的默认实现，其延后入口仍会同步执行动作。
     */
    internal val animationSeqHelper: DefaultAnimationSeqHelper
        get() = animationSeqHelperImpl ?: DefaultAnimationSeqHelper()

    /**
     * 返回进程共享配置容器，不创建副本，也不因工厂开关切换而替换配置实例。
     */
    internal val featureHelper: AnimationFeatureHelper
        get() = AnimationFeatureHelper

    /**
     * 把 Recents 清理请求交给当前启用的实际控制器，功能关闭时无操作。
     * 不会替换控制器或序列实例，也不接管宿主动画工厂；启用时遵守控制器的主线程约束。
     */
    internal fun cleanUpRecentsAnimation() {
        animationControllerImpl?.cleanUpRecentsAnim()
    }

    /**
     * 读取实际控制器是否存在；赋值必须在主线程且由 setter 锁串行化。
     * 启用时只补建缺失实例，重复启用保持身份；关闭时先摘除引用，再销毁旧控制器并确保清理旧序列任务。
     * 工厂开关不销毁独立配置订阅，能力声明也不会因此变为 false。
     */
    @PublicApi
    @set:Synchronized
    var interruptionEnabled: Boolean
        get() = animationControllerImpl != null
        set(enabled) {
            checkControllerMainThread()
            if (enabled) {
                if (animationControllerImpl == null) animationControllerImpl = AnimationController()
                if (animationSeqHelperImpl == null) animationSeqHelperImpl = AnimationSeqHelper()
            } else {
                val oldController = animationControllerImpl
                val oldSeq = animationSeqHelperImpl
                animationControllerImpl = null
                animationSeqHelperImpl = null
                try {
                    oldController?.destroy()
                } finally {
                    oldSeq?.clearFinishRecentsRunnable()
                }
            }
        }
}
